File: 019/main.js

 // Definition for singly-linked list.

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
function ListNode (val) {
  this.val = val
  this.next = null
}

var removeNthFromEnd = function (head, n) {
  let res = new ListNode(-1)
  res.next = head
  let mid = res
  let doc = head
  let arr = []
  while (doc !== null) {
    arr.push(doc.val)
    doc = doc.next
  }
  let len = arr.length
  let num = 0
  if (len === 1) return null
  while (res !== null) {
    if (num === len - n) {
      res.next = res.next.next
    }
    res = res.next
    num++
  }
  return mid.next
}

let l1 = {
  val: 1,
  next: {
    val: 2,
    next: {
      val: 3,
      next: {
        val: 4,
        next: {
          val: 5,
          next: null
        }
      }
    }
  }
}

let l2 = {
  val: 1,
  next: null
}

console.log(removeNthFromEnd(l1, 2))
console.log(removeNthFromEnd(l2, 1))

File: 019/main.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
 };

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* te = head;
        int num = 0;
        while (te) {
            te = te->next;
            num++;
        }
        int index = num - n;
        if (!index) {
            head = head->next;
        } else {
            te = head;
            while (--index) {
                te = te->next;
            }
            te->next = te->next->next;
        }
        return head;
    }
};

int main()
{
    Solution sol;
    ListNode* head = new ListNode(1);
    ListNode* head1 = new ListNode(2);
    ListNode* head2 = new ListNode(3);
    ListNode* head3 = new ListNode(4);
    ListNode* head4 = new ListNode(5);
    head3->next = head4;
    head2->next = head3;
    head1->next = head2;
    head->next = head1;
    ListNode* res = sol.removeNthFromEnd(head, 2);
    while (res) {
        cout << res->val << " ";
        res = res->next;
    }
    return 0;
}
